src/sdam/srv_polling/mod.rs

```diff
@@ -59,14 +59,21 @@ impl SrvPollingMonitor {
     }
 
     async fn execute(&mut self) {
-        while let Some(topology) = self.topology.upgrade() {
+        while self.topology.is_alive() {
+            let topology = match self.topology.upgrade() {
+                Some(topology) => topology,
+                None => break,
+            };
+
             let state = topology.clone_state().await;
 
             if state.is_sharded() || state.is_unknown() {
                 let hosts = self.lookup_hosts().await;
-                self.update_hosts(hosts, topology, state).await;
+                self.update_hosts(hosts, topology.clone(), state).await;
             }
 
+            std::mem::drop(topology);
```

Comment:
Unlike in the SDAM monitoring tasks, we don't ever need to wake up early due to messages from server selection, so we don't need the topology again until the next loop. As a slight optimization, we can drop the topology early and potentially help the other background tasks exit a little more quickly.